File: python/reboot-orchestrator/src/reboot_orchestrator/boot_state.py
"""
boot_state.py - Boot Identity Capture and Reboot Verification

ICMP reachability alone cannot prove that a host rebooted: a host that ignored
the reboot command (failed SSH authentication, denied sudo, hung shutdown unit)
keeps answering pings and looks indistinguishable from one that came back.

This module records a boot identity over SSH before the reboot is issued, reads
it again once the host is reachable, and compares the two to decide whether the
expected state change actually occurred.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Any
import subprocess
import time

from reboot_orchestrator.ssh import format_command, get_ssh_target_and_args

# Reads the two independent boot markers exposed by the kernel:
#   - /proc/sys/kernel/random/boot_id: regenerated on every boot (most reliable).
#   - /proc/uptime: seconds since boot, for hosts without a boot_id (busybox,
#     appliance firmware, network gear).
# Both are emitted as key=value lines and are empty when unreadable, so a host
# missing one marker still reports the other. The command deliberately contains
# no single quotes so that echoing it back to the operator stays readable.
BOOT_PROBE_COMMAND = (
    'printf "boot_id=%s\\n" "$(cat /proc/sys/kernel/random/boot_id 2>/dev/null)"; '
    'printf "uptime=%s\\n" "$(cut -d" " -f1 /proc/uptime 2>/dev/null)"'
)

# Slack allowed when comparing a post-reboot uptime against elapsed wall time,
# absorbing clock skew and the gap between issuing the probe and the host
# answering it.
UPTIME_TOLERANCE_SECONDS = 5.0


class VerificationStatus(Enum):
    """
    Outcome of comparing a host's boot identity before and after a reboot.
    """

    CONFIRMED = "confirmed"
    NOT_REBOOTED = "not-rebooted"
    UNKNOWN = "unknown"


@dataclass(frozen=True)
class BootState:
    """
    A point-in-time reading of a host's boot identity.

    Attributes:
        boot_id: Kernel boot UUID, or None when the host does not expose one.
        uptime_seconds: Seconds since boot, or None when unreadable.
        captured_at: Monotonic timestamp of the reading, used to measure the
                     wall-clock window a reboot had to occur in.
    """

    boot_id: str | None
    uptime_seconds: float | None
    captured_at: float

    @property
    def is_empty(self) -> bool:
        """
        Returns True when the host answered SSH but exposed no usable boot marker.
        """
        return self.boot_id is None and self.uptime_seconds is None


@dataclass(frozen=True)
class RebootVerification:
    """
    The verdict for a single host, suitable for printing and for deciding the
    process exit status.

    Attributes:
        host: Hostname the verdict applies to.
        status: Whether the reboot was confirmed, disproved, or indeterminate.
        detail: Human-readable evidence behind the verdict.
    """

    host: str
    status: VerificationStatus
    detail: str


def format_uptime(seconds: float) -> str:
    """
    Renders a duration in seconds as a compact human-readable uptime string.

    Args:
        seconds: Duration in seconds.

    Returns:
        str: Formatted duration such as "3d 4h 12m" or "45s".
    """
    total = int(seconds)
    days, rem = divmod(total, 86400)
    hours, rem = divmod(rem, 3600)
    minutes, secs = divmod(rem, 60)

    parts = []
    if days:
        parts.append(f"{days}d")
    if hours:
        parts.append(f"{hours}h")
    if minutes:
        parts.append(f"{minutes}m")
    if not parts:
        parts.append(f"{secs}s")

    return " ".join(parts)


def parse_boot_probe(output: str, captured_at: float) -> BootState:
    """
    Parses the key=value output of BOOT_PROBE_COMMAND into a BootState.

    Args:
        output: Raw stdout captured from the probe command.
        captured_at: Monotonic timestamp to record against the reading.

    Returns:
        BootState: Parsed markers; fields are None when absent or unparseable.
    """
    boot_id: str | None = None
    uptime: float | None = None

    for line in output.splitlines():
        key, _, value = line.partition("=")
        value = value.strip()
        if not value:
            continue
        if key.strip() == "boot_id":
            boot_id = value
        elif key.strip() == "uptime":
            try:
                uptime = float(value)
            except ValueError:
                continue

    return BootState(boot_id=boot_id, uptime_seconds=uptime, captured_at=captured_at)


def capture_boot_state(
    host: str,
    inventory: dict[str, dict[str, Any]],
    probe_timeout_seconds: int = 15,
    announce: bool = True,
) -> BootState | None:
    """
    Reads a host's boot identity over SSH.

    This doubles as an SSH pre-flight check: a host that cannot be probed almost
    certainly cannot be rebooted over SSH either.

    Args:
        host: Hostname of the target.
        inventory: Flattened inventory map of hosts to properties.
        probe_timeout_seconds: Seconds to wait for the SSH probe to complete.
        announce: Whether to print the command being run.

    Returns:
        BootState | None: The reading, or None when the host could not be probed.
    """
    target, ssh_args = get_ssh_target_and_args(host, inventory)
    cmd = ["ssh"] + ssh_args + [target, BOOT_PROBE_COMMAND]

    if announce:
        print(f"  Reading boot state of {host}:")
        print(f"    $ {format_command(cmd)}")

    try:
        res = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            check=False,
            timeout=probe_timeout_seconds,
        )
    except subprocess.TimeoutExpired:
        print(f"  WARNING: Boot state probe for '{host}' timed out.")
        return None
    except Exception as e:
        print(f"  WARNING: Boot state probe for '{host}' failed: {e}")
        return None

    captured_at = time.monotonic()

    if res.returncode != 0:
        print(
            f"  WARNING: Boot state probe for '{host}' returned code {res.returncode}. "
            f"Error: {res.stderr.strip()}"
        )
        return None

    state = parse_boot_probe(res.stdout, captured_at)
    if state.is_empty:
        print(
            f"  WARNING: '{host}' exposed neither /proc/sys/kernel/random/boot_id "
            "nor /proc/uptime; its reboot cannot be verified."
        )
    return state


def verify_reboot(
    host: str,
    before: BootState | None,
    after: BootState | None,
) -> RebootVerification:
    """
    Compares boot identities recorded before and after a reboot.

    A changed boot_id is definitive. Failing that, uptime is used: a host that
    rebooted must report an uptime lower than its previous one and no larger
    than the wall-clock window between the two readings.

    Args:
        host: Hostname the readings belong to.
        before: Reading taken before the reboot was issued, or None.
        after: Reading taken once the host was reachable again, or None.

    Returns:
        RebootVerification: The verdict and the evidence behind it.
    """
    if before is None and after is None:
        return RebootVerification(
            host,
            VerificationStatus.UNKNOWN,
            "boot state could not be read before or after the reboot",
        )
    if before is None:
        return RebootVerification(
            host,
            VerificationStatus.UNKNOWN,
            "no baseline boot state was recorded before the reboot",
        )
    if after is None:
        return RebootVerification(
            host,
            VerificationStatus.UNKNOWN,
            "host answered ping but its boot state could not be read afterwards",
        )

    if before.boot_id and after.boot_id:
        if before.boot_id != after.boot_id:
            return RebootVerification(
                host,
                VerificationStatus.CONFIRMED,
                f"boot_id changed ({before.boot_id} -> {after.boot_id})",
            )
        return RebootVerification(
            host,
            VerificationStatus.NOT_REBOOTED,
            f"boot_id is unchanged ({before.boot_id}); the host never went down",
        )

    if before.uptime_seconds is not None and after.uptime_seconds is not None:
        elapsed = after.captured_at - before.captured_at
        before_str = format_uptime(before.uptime_seconds)
        after_str = format_uptime(after.uptime_seconds)
        if (
            after.uptime_seconds < before.uptime_seconds
            or after.uptime_seconds <= elapsed + UPTIME_TOLERANCE_SECONDS
        ):
            return RebootVerification(
                host,
                VerificationStatus.CONFIRMED,
                f"uptime reset ({before_str} -> {after_str})",
            )
        return RebootVerification(
            host,
            VerificationStatus.NOT_REBOOTED,
            f"uptime kept climbing ({before_str} -> {after_str}) over a "
            f"{format_uptime(elapsed)} window; the host never went down",
        )

    return RebootVerification(
        host,
        VerificationStatus.UNKNOWN,
        "host exposes no boot_id or uptime marker to compare",
    )

File: go/pkg/reboot/bootstate.go
package reboot

import (
	"context"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"
)

// bootProbeCommand reads the two independent boot markers the kernel exposes:
//
//   - /proc/sys/kernel/random/boot_id, regenerated on every boot, and so
//     authoritative.
//   - /proc/uptime, the fallback for hosts exposing no boot ID at all —
//     busybox-based firmware, appliances, network gear.
//
// Both are emitted as key=value lines and left empty when unreadable, so a host
// missing one marker still reports the other. The command deliberately contains
// no single quotes, keeping it readable when echoed back to the operator.
const bootProbeCommand = `printf "boot_id=%s\n" "$(cat /proc/sys/kernel/random/boot_id 2>/dev/null)"; ` +
	`printf "uptime=%s\n" "$(cut -d" " -f1 /proc/uptime 2>/dev/null)"`

// uptimeTolerance absorbs clock skew and the gap between issuing a probe and
// the host answering it when comparing uptime against elapsed wall time.
const uptimeTolerance = 5 * time.Second

// VerificationStatus is the outcome of comparing a host's boot identity before
// and after a reboot.
type VerificationStatus int

const (
	// StatusConfirmed means the host provably restarted.
	StatusConfirmed VerificationStatus = iota
	// StatusNotRebooted means the host provably did not restart.
	StatusNotRebooted
	// StatusUnknown means the evidence does not settle it either way.
	StatusUnknown
)

// String renders the status for display.
func (s VerificationStatus) String() string {
	switch s {
	case StatusConfirmed:
		return "confirmed"
	case StatusNotRebooted:
		return "not-rebooted"
	default:
		return "unknown"
	}
}

// BootState is a point-in-time reading of a host's boot identity.
type BootState struct {
	// BootID is the kernel boot UUID, empty when the host exposes none.
	BootID string
	// Uptime is the time since boot; zero when unreadable, which HasUptime
	// distinguishes from a genuine reading.
	Uptime time.Duration
	// HasUptime reports whether Uptime was actually read.
	HasUptime bool
	// CapturedAt is when the reading was taken, used to measure the wall-clock
	// window a reboot had to happen in.
	CapturedAt time.Time
}

// isEmpty reports whether the host answered but exposed no usable marker.
func (b BootState) isEmpty() bool {
	return b.BootID == "" && !b.HasUptime
}

// RebootVerification is the verdict for a single host.
type RebootVerification struct {
	// Host is the host the verdict applies to.
	Host string
	// Status is whether the reboot was confirmed, disproved, or indeterminate.
	Status VerificationStatus
	// Detail is the human-readable evidence behind the verdict.
	Detail string
}

// parseBootProbe parses the key=value output of bootProbeCommand.
func parseBootProbe(output string, capturedAt time.Time) BootState {
	state := BootState{CapturedAt: capturedAt}
	for _, line := range strings.Split(output, "\n") {
		key, value, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		value = strings.TrimSpace(value)
		if value == "" {
			continue
		}
		switch strings.TrimSpace(key) {
		case "boot_id":
			state.BootID = value
		case "uptime":
			seconds, err := strconv.ParseFloat(value, 64)
			if err != nil {
				continue
			}
			state.Uptime = time.Duration(seconds * float64(time.Second))
			state.HasUptime = true
		}
	}
	return state
}

// CaptureBootState reads a host's boot identity over SSH.
//
// This doubles as an SSH pre-flight check: a host that cannot be probed almost
// certainly cannot be rebooted over SSH either. A nil return means the host
// could not be read at all, which is deliberately distinct from a reading that
// contained no markers.
func CaptureBootState(
	ctx context.Context,
	out io.Writer,
	runner Runner,
	clock Clock,
	host Host,
	timeout time.Duration,
) *BootState {
	// The echoed command is the whole announcement: it names the host, and the
	// printf it carries says plainly what is being read.
	args := sshCommand(host, bootProbeCommand)
	reportHost(out, host.Name, "$ %s", FormatCommand("ssh", args...))

	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	stdout, err := runner.Run(ctx, "", "ssh", args...)
	if err != nil {
		reportHost(out, host.Name, "WARNING: boot state probe failed: %v", err)
		return nil
	}

	state := parseBootProbe(stdout, clock.Now())
	if state.isEmpty() {
		reportHost(out, host.Name, "WARNING: exposed neither /proc/sys/kernel/random/boot_id "+
			"nor /proc/uptime; its reboot cannot be verified.")
	}
	return &state
}

// VerifyReboot decides whether a host restarted, from its boot markers first
// and the monitor's observation second.
//
// Boot markers stay authoritative: they are read from the host itself and say
// what happened rather than what was seen from outside. The observed power
// cycle only breaks ties the markers leave open — which is most valuable on
// hosts that expose no marker at all, whose reboots were previously
// unverifiable in either direction.
func VerifyReboot(host string, before, after *BootState, cycle Cycle) RebootVerification {
	verdict := verifyByMarkers(host, before, after)
	if verdict.Status != StatusUnknown {
		return verdict
	}
	switch {
	case cycle.Complete():
		return RebootVerification{host, StatusConfirmed, fmt.Sprintf(
			"%s, but it was seen to go down at %s and come back %s later",
			verdict.Detail, stamp(cycle.DownAt), formatUptime(cycle.DownFor()))}
	case cycle.StayedUp():
		return RebootVerification{host, StatusNotRebooted, fmt.Sprintf(
			"%s, and it answered every probe throughout; it never went down", verdict.Detail)}
	}
	return verdict
}

// verifyByMarkers compares boot identities recorded before and after a reboot.
//
// A changed boot ID is definitive. Failing that, uptime decides: a host that
// restarted must report an uptime lower than its previous one, or one no larger
// than the window between the two readings.
func verifyByMarkers(host string, before, after *BootState) RebootVerification {
	switch {
	case before == nil && after == nil:
		return RebootVerification{host, StatusUnknown,
			"boot state could not be read before or after the reboot"}
	case before == nil:
		return RebootVerification{host, StatusUnknown,
			"no baseline boot state was recorded before the reboot"}
	case after == nil:
		return RebootVerification{host, StatusUnknown,
			"host answered ping but its boot state could not be read afterwards"}
	}

	if before.BootID != "" && after.BootID != "" {
		if before.BootID != after.BootID {
			return RebootVerification{host, StatusConfirmed,
				fmt.Sprintf("boot_id changed (%s -> %s)", before.BootID, after.BootID)}
		}
		return RebootVerification{host, StatusNotRebooted,
			fmt.Sprintf("boot_id is unchanged (%s); the host never went down", before.BootID)}
	}

	if before.HasUptime && after.HasUptime {
		elapsed := after.CapturedAt.Sub(before.CapturedAt)
		beforeStr, afterStr := formatUptime(before.Uptime), formatUptime(after.Uptime)
		if after.Uptime < before.Uptime || after.Uptime <= elapsed+uptimeTolerance {
			return RebootVerification{host, StatusConfirmed,
				fmt.Sprintf("uptime reset (%s -> %s)", beforeStr, afterStr)}
		}
		return RebootVerification{host, StatusNotRebooted,
			fmt.Sprintf("uptime kept climbing (%s -> %s) over a %s window; the host never went down",
				beforeStr, afterStr, formatUptime(elapsed))}
	}

	return RebootVerification{host, StatusUnknown,
		"host exposes no boot_id or uptime marker to compare"}
}

// formatUptime renders a duration as a compact human-readable string.
func formatUptime(d time.Duration) string {
	total := int(d.Seconds())
	if total < 0 {
		total = 0
	}
	days, rem := total/86400, total%86400
	hours, rem := rem/3600, rem%3600
	minutes, seconds := rem/60, rem%60

	var parts []string
	if days > 0 {
		parts = append(parts, fmt.Sprintf("%dd", days))
	}
	if hours > 0 {
		parts = append(parts, fmt.Sprintf("%dh", hours))
	}
	if minutes > 0 {
		parts = append(parts, fmt.Sprintf("%dm", minutes))
	}
	if len(parts) == 0 {
		parts = append(parts, fmt.Sprintf("%ds", seconds))
	}
	return strings.Join(parts, " ")
}
